### ISSUE TITLE
Typescript examples

### ISSUE THREAD

**Issue by ashishsc** [redux-loop/redux-loop#160]
I think we should try writing the example in typescript. There's some friction you run into when defining action creators unless you define it the way redux-loop is expecting it (using the ActionCreator type, etc.) I'll try and post a concrete issue of what I'm facing here in a bit. @willisplummer since you added the definitions, do you have any sample code using redux-loop in a ts project?

**Comment by bdwain** [comment]
Agreed. This is also related to #158. I think that's probably the next big task ahead of us. It would really lower the barrier to entry for people.

**Comment by ashishsc** [comment]
Ok so, first a question about the type for Cmd.run:
```typescript
  static readonly run: <A extends Action>(
    f: Function,
    options?: {
      args?: any[];
      failActionCreator?: ActionCreator<A>;
      successActionCreator?: ActionCreator<A>;
      forceSync?: boolean;
    }
  ) => RunCmd<A>;
```

I'm getting type errors here because my fail and success action don't create the same action. Shouldn't the actual type be
```typescript
  static readonly run: <A extends Action, F extends Action>(
    f: Function,
    options?: {
      args?: any[];
      failActionCreator?: ActionCreator<F>;
      successActionCreator?: ActionCreator<A>;
      forceSync?: boolean;
    }
  ) => RunCmd<A>;
```

?

**Comment by ashishsc** [comment]
Another issue:
```typescript
export type ReducerMapObject<S, A extends Action = AnyAction> = {
    [K in keyof S]: LoopReducer<S[K], A>
};

declare function combineReducers<S, A extends Action = AnyAction>(
    reducers: ReducerMapObject<S, A>,
): LiftedLoopReducer<S, A>;

declare function mergeChildReducers<S, A extends Action = AnyAction>(
    parentResult: S | Loop<S, A>,
    action: AnyAction,
    childMap: ReducerMapObject<S, A>,
): Loop<S, A>;
```

including redux-loop in my project gives me errors:
```
ERROR at [repo]/node_modules/redux-loop/index.d.ts(123,50): 
TS1005: ',' expected.
ERROR at [repo]i/node_modules/redux-loop/index.d.ts(127,54): 
TS1005: ',' expected.
ERROR at [repo]/node_modules/redux-loop/index.d.ts(128,13): 
TS2314: Generic type 'ReducerMapObject' requires 3 type argument(s).
ERROR at [repo]/node_modules/redux-loop/index.d.ts(131,57): 
TS1005: ',' expected.
ERROR at [repo]/node_modules/redux-loop/index.d.ts(134,13): 
TS2314: Generic type 'ReducerMapObject' requires 3 type argument(s).
```

**Comment by bdwain** [comment]
Having played around with typescript a bit to work with the typings, I feel like rewriting everything with typescript might make it harder for people who don't use typescript to contribute. That, combined with the effort it will takes makes me think it's probably not worth it. 

I think we could improve our tests to catch issues with our typings though for sure. We really need to improve the examples folder and a few different kinds in there, including a typescript project. I think that'd be really helpful.

**Comment by gniquil** [comment]
@ashishsc For you first issue, `Cmd.run`, I've run into this as well. The solution is to make your action creators return the union `Action` type. That should solve your problem. However, if you want to adopt this approach, your action creators need to be able to import the root Action type, which implies that action creators need to exist completely separately from your action definitions. This is generally different from the way people code in normal javascript (say duck style). 

Step back a bit, and being a super fan of Elm (but not be able to use it for various reasons), adopting this library with typescript, you should stick to the Elm architecture as much as possible. Strict typing points out many issues exist in the normal redux architecture. So, decompose your app into multiple sub reducers and use `Cmd.map` to wrap actions (generally the "fractal pattern"). At work, once we started doing this, most of our typing related issues are gone.

**Comment by ashishsc** [comment]
@gniquil I am also a huge fan of elm, hence why I came to loop. I am not sure that our action creators should return the union ```Action``` type. I take advantage of the fact that the creators can return a specific kind of action. Although now that you mention it, that's similar to how in elm, the view functions return ```elm Html Msg```  where ```Msg``` is a union type.  @willisplummer was that the intent?

@bdwain I don't it would be too bad to rewrite in typescript since we already have the typings written, we just need to annotate our actual functions with them, but I do agree that it would make it harder for non-tsers to contribute. I also agree that we really need an example of using typescript in the examples folder. I could take that on after I get this working.

**Comment by willisplummer** [comment]
Yeah, @gniquil has the right idea here. Take a look at the tests i wrote for the typescript definitions to get a sense of how this is working: https://github.com/redux-loop/redux-loop/blob/master/test/typescript/loopReducer.ts

I agree that a TypeScript rewrite wouldn't actually be very hard to do. Adding function annotations is generally fairly straightforward, and it would give us more confidence that our types were accurate. It's easy for a declaration file to get out of sync with the JavaScript code if someone makes a change and doesn't remember to update the types, and actually using `.ts` files would prevent this possibility.

The `ReducerMapObject` errors you're running into are weird to me. AFAICT the compiler should be able to infer the type of `K` from `S`, right? (cc @jaredklewis )

**Comment by willisplummer** [comment]
@ashishsc I'm not sure what an action creator would return if not the union type of all the possible actions for a given reducer or more broadly redux's `AnyAction` type?

**Comment by willisplummer** [comment]
Another argument that I see for pursuing a TypeScript rewrite is that defining your API's can quickly make architectural complexity more apparent. (When a function accepts weird union types or is difficult to type, that's generally a good indicator that it's attempting to do too much).

**Comment by gniquil** [comment]
Like `Cmd.run`. Wouldn't it be awesome if it type checks the `args` against effect func arguments, and resolved promise wrapper against success/fail actionCreators. Too bad typescript can't do the following

```js
Cmd.run: <T, R, A>(f: (...args: T) => Promise<R>, { args: T, successActionCreator: (result: R) => A, failActionCreator: (error: any) => A }): CmdType<...>
```

At work we reexport a more restricted version of Cmd.run, forcing effect func to only take 1 arg (and completely forbid getState and dispatch).

**Comment by ashishsc** [comment]
I take issue with the way the Redux uses AnyAction, I like to do in my reducers and subreducers
```javascript
switch(action.type)
    case FOO: {
         ....
    },
    ...,
    default:
        thisFailsAtCompileTime(action)
```

but with the way redux defines AnyAction this isn't possible.

**Comment by ashishsc** [comment]
Wow I'm so blind. Sorry guys I didn't even see https://github.com/redux-loop/redux-loop/blob/master/test/typescript/loopReducer.ts . It'd be worth moving that into the examples dir.

Also I think my type errors could be a tsc version mismatch? I'll confirm later.

**Comment by bdwain** [comment]
@ashishsc I think the typescript test is good for a test, but it's kind of just a bunch of random stuff.  I think we should have a full fledged app example in the examples folder. Maybe that would also make a good test too.

My two main issues with making the whole library typescript is that it
- Makes it harder for people who don't use typescript to contribute, while typescript users can easily contribute to a js library. Considering not many other redux/react libraries use typescript, I'm worried we'd cut off a lot of potential contributors. 
- Signals to users of the library that typescript is encouraged and the ideal way to use the library, which also would turn people off (that's what happened with angular 2 for a lot of people)

If the community started moving more in the direction of typescript, I think we'd see other libraries (especially redux) start doing the same thing and then I think those two issues would be much less of a concern.

For now, is it alright if we just work to improve the reliability of our typings using tests and examples? If that still ends up being a problem, then maybe converting the whole thing would be the answer. I'd rather take the least drastic step we need to have reliable typings for people using the library.

**Comment by ashishsc** [comment]
I think angular2's main issue was the confusing tooling, identity-crisis api, and several other issues indirectly related or maybe slightly compounded by typescript. I do agree that we shoul dhave an example for now though, I wouldn't want to force ts or make it required.

**Comment by ashishsc** [comment]
So I confirmed that my issues were from using typescript@2.4, the issues went away after, I should've realized that sooner. This thread can keep serving as a discussion for what we want to do with typescript, however.

**Comment by gniquil** [comment]
@willisplummer Looking forward to seeing your PR. But just now I came across this:

```js
function f<...T,...U>() {}
}
class C<...T> {
}
```
https://github.com/Microsoft/TypeScript/issues/5453

Variadic kinds in 2.7. It would solve so many problems...

**Comment by gniquil** [comment]
Oops mis-spoke, not 2.7, but at least in the roadmap.

https://github.com/Microsoft/TypeScript/wiki/Roadmap

**Comment by seethroughtrees** [comment]
I appear to be having the same type issue as @ashishsc for Cmd. run in https://github.com/redux-loop/redux-loop/issues/160#issuecomment-335519497, where the output is expecting all actions to have the same type.  

i'm not sure how if the example linked in here explains how this is any different?  can someone elaborate on how we should type the output of loops that use Cmd.run ?

**Comment by willisplummer** [comment]
@seethroughtrees if you create a union of all the possible actions and make that the output for your reducer, that should solve the issue?

**Comment by seethroughtrees** [comment]
It seems your response was deleted @willisplummer , but thanks for the response and the typings you provided for the library!  

In case anyone has this problem in the future, I found typing the output of the individual action creators was creating the conflict for me.  I changed each action creator output type to the same generic union type as I'm passing to the reducer.  Not sure if this is ideal, but it now works as expected.   Thanks again.

**Comment by willisplummer** [comment]
@seethroughtrees glad to hear you got it working - sounds like you're doing things correctly. i wrote that comment out and then deleted because i wanted to revisit when i had a little more time to articulate how it should be done 😬

if you wanted to help out, there have been a few requests for additional examples and documentation of how to use Loop w/ Typescript

**Comment by ashishsc** [comment]
@seethroughtrees That is in fact how it is meant to be used. I didn't realize it at first, but it is the same as the way Elm defines its Cmds in that respect. The benefit of making it generic to the union Action type in the actions is that it allows typescript to do the work to figure out which action is being called, letting you change which one of your specific actions you want to call without having to update the signature. This makes sense because in practice, you generally want your actions to be called within the same scope (in other words, same set of actions relevant to your module.)

Sorry if I'm not articulating that well.

**Comment by seethroughtrees** [comment]
Makes perfect sense @ashishsc , thanks for the perspective.  

@willisplummer , i can definitely help there.   this is my first project using typescript instead of flow.  when i get my patterns finalized, i'll be happy to submit some examples.  :)

**Comment by barberousse** [comment]
Another thing about y'alls types is that this:
```typescript
export interface LoopReducer<S, A extends Action> {
  (state: S | undefined, action: AnyAction): S | Loop<S, A>;
}
```

is really bad in that a reducer should never return `undefined` nor should the state parameter ever be `undefined`. You have to realize you're forcing everyone downstream to type their Reducer to expect undefined as well insert type guards everywhere to satisfy `tsc`.

**Comment by bdwain** [comment]
@barberousse Redux will pass undefined to a reducer initially so that you can return your initial state.

> We'll start by specifying the initial state. Redux will call our reducer with an undefined state for the first time. This is our chance to return the initial state of our app:

(from [redux docs](https://redux.js.org/basics/reducers#handling-actions))

**Comment by barberousse** [comment]
It does, but it's an implementation detail that shouldn't be exposed to the
type system.

**Comment by bdwain** [comment]
How can a reducer accept an undefined state if it's explicitly typed to not accept it?

**Comment by bdwain** [comment]
[Redux's types do the same thing](https://github.com/reactjs/redux/blob/master/index.d.ts#L59)

**Comment by barberousse** [comment]
It's a non-issue. The type system only goes as far as compile time, but you
don't write your reducer code against `undefined` anyway. The downside of
leaking that into the type aglebra is that your reducer return values all
have to be passed through a type guard. If I have some state tree, each
leaf is forced to either be the union with `undefined` or you pass the
reducer result through a type guard. You'd eliminate a lot of needless
boiler plate if you simply didn't expose that implementation detail. Lemme
know if there's something I'm missing.

**Comment by bdwain** [comment]
```js
export interface LoopReducer<S, A extends Action> {
  (state: S | undefined, action: AnyAction): S | Loop<S, A>;
}
```
The return value does not allow for an undefined return value as far as I can tell. Only a state or a loop. Undefined is only mentioned as an input type for the state input.

Also, if the reducer was not typed to allow undefined as a State value, then this bad code would get through the type checking.

```js
//no default param for initialState
function reducer(state, action){
   switch(action.type){ //pretty sure there is no action param when redux passes undefined
      //handle some action types
     default:
        return state; //would return undefined if state was undefined
   }
}
```

That would break immediately at runtime but be valid according to the type system. If you instead say that undefined is a valid input for parameter 1, but not a valid return value, it would detect that you tried to return a value that was potentially undefined and fail typechecking.

**Comment by barberousse** [comment]
That's fair

**Comment by bdwain** [comment]
I'm going to close this since there's an example on the site now.

**Comment by RichieAHB** [comment]
As an addendum to this I'm using `reduceReducers` to bolt on a side effecting loop reducer that has access to the whole state tree (using selectors to access what it needs). I had the above issue with typescript complaining about `undefined`. I figured that to get the initial state from my plain reducer I'd just do what redux does under the hood and call the reducer for the first time:

```js
const initialState = root(undefined, { type: '@@@@@' });

const effects = (
  state: State = initialState,
  action: Action
): State | Loop<State, Action> => { /* ... */ }
```

It won't break at runtime and will keep up to date with any changes in my other reducers 👍 - thought it was worth sharing as it seems obvious but the fact I'm on this issue suggests I had to do some searching to get around this issue myself!
